### ISSUE TITLE
Mounting /dev inside a container breaks /dev/ptmx for the host

### ISSUE THREAD

**Issue by mortenlj** [docker/docker#19464]
I'm following the guide for setting up kubernetes on my laptop for testing (http://kubernetes.io/v1.1/docs/getting-started-guides/docker.html), and when I do the second step (starting the master) my host computer encounters a number of problems.

After much digging, it comes down to /dev/ptmx being substituted with a symlink to /dev/pts/ptmx, which does not have the same access rights (basically no user on the host is allowed to access it). This causes all sorts of problems for various terminal emulators.

See transcript below:

```
┌─[fimojoha@fimojoha-w530]─[~]
└──╼ ll /dev/ptmx 
crw-rw-rw- 1 root tty 5, 2 jan.  19 22:25 /dev/ptmx
┌─[fimojoha@fimojoha-w530]─[~]
└──╼ docker run \
>    --volume=/dev:/dev \
>     --net=host \
>     --pid=host \
>     --privileged=true \
>     -d \
>     gcr.io/google_containers/hyperkube:v1.0.1 \
>     /hyperkube kubelet --containerized --hostname-override="127.0.0.1" --address="0.0.0.0" \
>     --api-servers=http://localhost:8080 --config=/etc/kubernetes/manifests
fb9e98f2c34b96149a8951ac837ba20645a424e57e4e7c2233850aea02117f42
┌─[fimojoha@fimojoha-w530]─[~]
└──╼ ll /dev/ptmx 
lrwxrwxrwx 1 root root 8 jan.  19 22:27 /dev/ptmx -> pts/ptmx
┌─[fimojoha@fimojoha-w530]─[~]
└──╼ ll /dev/pts/ptmx 
c--------- 1 root root 5, 2 jan.  19 22:25 /dev/pts/ptmx
```

Additional info:

```
┌─[fimojoha@fimojoha-w530]─[~]
└──╼ docker version
Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        
 OS/Arch:      linux/amd64
┌─[fimojoha@fimojoha-w530]─[~]
└──╼ docker info
Containers: 52
Images: 79
Server Version: 1.9.1
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.1.12-gentoo
Operating System: <unknown>
CPUs: 8
Total Memory: 15.32 GiB
Name: fimojoha-w530
ID: F6JQ:YD4I:LMHW:T7B5:T3E4:QLIV:GFRE:72MY:CYCC:ESBU:NE5K:EFOT
Username: mortenlj
Registry: https://index.docker.io/v1/
```

**Comment by HackToday** [comment]
Hi, I am not sure why you hit that in docker 1.9.1.

I just tried on docker RC build, it has following output:

tried on scheduler and controller,

```
ubuntu@speedwork:~$ docker exec -it 923518b7000f ls -l /dev/ptmx
lrwxrwxrwx 1 root root 8 Jan 20 06:58 /dev/ptmx -> pts/ptmx
ubuntu@speedwork:~$ docker exec -it 923518b7000f ls -l /dev/pts/ptmx
crw-rw-rw- 1 root root 5, 2 Jan 20 06:58 /dev/pts/ptmx
ubuntu@speedwork:~$ docker exec -it 44434758c58b ls -l /dev/ptmx
lrwxrwxrwx 1 root root 8 Jan 20 06:58 /dev/ptmx -> pts/ptmx
ubuntu@speedwork:~$ docker exec -it 44434758c58b ls -l /dev/pts/ptmx
crw-rw-rw- 1 root root 5, 2 Jan 20 06:58 /dev/pts/ptmx
```

Seems ok.  Which kubernetes container do you have issue ?

**Comment by mortenlj** [comment]
You can see the command/container in the initial report, where I stripped out a few options that didn't affect the issue, but just copying the command in step two of the guide I linked to is enough to trigger this problem:

```
docker run \
    --volume=/:/rootfs:ro \
    --volume=/sys:/sys:ro \
    --volume=/dev:/dev \
    --volume=/var/lib/docker/:/var/lib/docker:ro \
    --volume=/var/lib/kubelet/:/var/lib/kubelet:rw \
    --volume=/var/run:/var/run:rw \
    --net=host \
    --pid=host \
    --privileged=true \
    -d \
    gcr.io/google_containers/hyperkube:v1.0.1 \
    /hyperkube kubelet --containerized --hostname-override="127.0.0.1" --address="0.0.0.0" \
    --api-servers=http://localhost:8080 --config=/etc/kubernetes/manifests
```

Also, note that the problem is outside docker, on the *host* computer, not inside the container. When mounting /dev inside the container, docker makes changes to /dev on the *host* that breaks terminal emulators running on the *host*.

**Comment by HackToday** [comment]
Still not reproduce this issue according to your comment above

```
ubuntu@speedwork:~$ ls -l /dev/ptmx 
crw-rw-rw- 1 root tty 5, 2 Jan 20 09:21 /dev/ptmx
ubuntu@speedwork:~$ 
ubuntu@speedwork:~$ docker run \
>     --volume=/:/rootfs:ro \
>     --volume=/sys:/sys:ro \
>     --volume=/dev:/dev \
>     --volume=/var/lib/docker/:/var/lib/docker:ro \
>     --volume=/var/lib/kubelet/:/var/lib/kubelet:rw \
>     --volume=/var/run:/var/run:rw \
>     --net=host \
>     --pid=host \
>     --privileged=true \
>     -d \
>     gcr.io/google_containers/hyperkube:v1.0.1 \
>     /hyperkube kubelet --containerized --hostname-override="127.0.0.1" --address="0.0.0.0" \
>     --api-servers=http://localhost:8080 --config=/etc/kubernetes/manifests
25e6a619c74aa237f1eacfb987f0070e9ebf55fbf1aa1a140cfcb9c3b575d20b
ubuntu@speedwork:~$ ls -l /dev/ptmx 
crw-rw-rw- 1 root tty 5, 2 Jan 20 09:21 /dev/ptmx



```

**Comment by mortenlj** [comment]
Any idea about what could be different on my setup, which could cause this to happen? I don't know enough about how /dev and /dev/pts works to know what to look for.

It consistenly happens on my machine whenever I run that command.

**Comment by HackToday** [comment]
maybe enable docker daemon -D logs can show more valuable message,  Also you can check more about in dmesg if anything can help identify

**Comment by ibuildthecloud** [comment]
I've ran into this in the past.  Do not bind mount `/dev:/dev`, bad bad bad things can happen.  It would be nice if libcontainer could recognized this situation but I think it's rather hard.  If you need the host /dev you should really do `/dev:/host/dev` and then in the entrypoint do `mount --rbind /host/dev /dev`

**Comment by phemmer** [comment]
@ibuildthecloud bind mounting `/dev:/dev` works just fine. Docker fully supports it. See #16639

**Comment by mortenlj** [comment]
@ibuildthecloud I don't think I can do that, because I'm not running an image of my own making, but using an image from kubernetes. I'm following their instructions to the letter and getting this problem.

@phemmer I'm trying to figure out if that change is part of docker 1.9.1, or if it's not yet released, but I can't find it in any changelogs. From the comments on that PR, it looks like it should have prevented me from getting this problem, but if it's part of 1.9.1, there's something missing.

@HackToday I tried enabling -D, but I can't see anything in the logs that looks like it could be relevant. There's nothing in dmesg about what's happening.

**Comment by phemmer** [comment]
@mortenlj it's not part of 1.9.1. It'll be in 1.10.
If you want to know how to determine this yourself, go to the PR, and click the merge commit hash (the bit where where it says "*LK4D4 merged commit 9187656 ...*"). At the top, in the blue section it'll show you a list of all the branches and tags that commit is present in. Notice right now the only tag is `v1.10.0-rc1`.

**Comment by mortenlj** [comment]
Thanks. Then it looks like this issue will be resolved by docker 1.10 when that is released.
Until then I have a workaround with chmod'ing /dev/pts/ptmx after starting the container.

I'll close this issue now. Thanks for the help.
